library/Control/Monad/AWS/Matchers.hs

```diff
@@ -0,0 +1,87 @@
+module Control.Monad.AWS.Matchers
+  ( Matchers
+  , HasMatchers (..)
+  , Matcher (..)
+  , withMatcher
+  , withMatchers
+  , matchSend
+  , matchAwait
+  ) where
+
+import Prelude
+
+import Amazonka (AWSRequest, AWSResponse, Error)
+import qualified Amazonka.Waiter as Waiter
+import Control.Lens (Lens', view, (<>~))
+import Control.Monad (guard)
+import Control.Monad.Reader (MonadReader (..))
+import Data.Maybe (listToMaybe, mapMaybe)
+import Data.Typeable
+import GHC.Stack (HasCallStack)
+
+-- | Define a response to provide for any matched requests
+data Matcher where
+  -- Matches calls to 'send' where the given predicate holds
+  SendMatcher
+    :: forall a
+     . (AWSRequest a, Typeable a, Typeable (AWSResponse a))
+    => (a -> Bool)
+    -> Either Error (AWSResponse a)
+    -> Matcher
+  -- Matches calls to 'await' where the given predicate holds
+  AwaitMatcher
+    :: forall a
+     . (AWSRequest a, Typeable a)
+    => (Waiter.Wait a -> a -> Bool)
+    -> Either Error Waiter.Accept
+    -> Matcher
+
+newtype Matchers = Matchers
+  { unMatchers :: [Matcher]
+  }
+  deriving newtype (Semigroup, Monoid)
+
+class HasMatchers env where
+  matchersL :: Lens' env Matchers
+
+instance HasMatchers Matchers where
+  matchersL = id
+
+-- | Add a 'Matcher' for the duration of the block
+withMatcher :: (MonadReader env m, HasMatchers env) => Matcher -> m a -> m a
+withMatcher = withMatchers . pure
+
+-- | Add multiple 'Matcher's for the duration of the block
+withMatchers :: (MonadReader env m, HasMatchers env) => [Matcher] -> m a -> m a
+withMatchers ms = local $ matchersL <>~ Matchers ms
+
+matchSend
+  :: (MonadReader env m, HasMatchers env, Typeable a, Typeable (AWSResponse a))
+  => a
+  -> m (Either Error (AWSResponse a))
+matchSend req = firstMatcher $ \case
+  SendMatcher matchReq resp -> do
+    guard . matchReq =<< cast req
+    cast resp
+  AwaitMatcher {} -> Nothing
+
+matchAwait
+  :: (MonadReader env m, HasMatchers env, Typeable a)
+  => Waiter.Wait a
+  -> a
+  -> m (Either Error Waiter.Accept)
+matchAwait w req = firstMatcher $ \case
+  SendMatcher {} -> Nothing
+  AwaitMatcher matchReq acc -> do
+    guard =<< matchReq <$> cast w <*> cast req
+    cast acc
+
+firstMatcher
+  :: (HasCallStack, MonadReader env m, HasMatchers env)
+  => (Matcher -> Maybe a)
+  -> m a
+firstMatcher f = do
+  matchers <- view matchersL
+  maybe errUnmatched pure $ listToMaybe $ mapMaybe f $ unMatchers matchers
+ where
+  errUnmatched = error "TODO"
```

Comment:
I'd fill this in here, alert that no matcher matched a given call.

Reply:
not to jump on you in case you already knew this needed to be done - I just came looking to see what happened if a `send` did not get matched, and saw this.